server/src/main/java/com/linecorp/centraldogma/server/internal/api/GitHttpService.java

```diff
@@ -0,0 +1,204 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.common.AggregatedHttpRequest;
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.ServerCacheControl;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.stream.ByteStreamMessage;
+import com.linecorp.armeria.common.stream.StreamMessage;
+import com.linecorp.armeria.server.annotation.Get;
+import com.linecorp.armeria.server.annotation.Header;
+import com.linecorp.armeria.server.annotation.Param;
+import com.linecorp.armeria.server.annotation.Post;
+import com.linecorp.centraldogma.server.internal.api.auth.RequiresReadPermission;
+import com.linecorp.centraldogma.server.internal.storage.project.ProjectApiManager;
+
+/**
+ * A service that provides Git HTTP protocol.
+ */
+@RequiresReadPermission
+public final class GitHttpService {
+
+    private static final Logger logger = LoggerFactory.getLogger(GitHttpService.class);
+
+    private static final char[] HEX_DIGITS = "0123456789abcdef".toCharArray();
+
+    // TODO(minwoox): Add the headers in this class to Armeria.
+    private static final AggregatedHttpResponse CAPABILITY_ADVERTISEMENT_RESPONSE = AggregatedHttpResponse.of(
+            ResponseHeaders.builder(200)
+                           .add(HttpHeaderNames.CONTENT_TYPE, "application/x-git-upload-pack-advertisement")
+                           .add(HttpHeaderNames.CACHE_CONTROL, ServerCacheControl.REVALIDATED.asHeaderValue())
+                           .build(),
+            HttpData.ofUtf8(capabilityAdvertisement()));
+
+    // https://git-scm.com/docs/protocol-capabilities/
+    private static String capabilityAdvertisement() {
+        final StringBuilder sb = new StringBuilder();
+        sb.append("001e# service=git-upload-pack\n");
+        pktFlush(sb);
+        pktLine(sb, "version 2");
+        pktLine(sb, COMMAND_LS_REFS);
+        // Support limited options for now due to the unique characteristics of Git repositories in
+        // Central Dogma, such as having only a master branch and no tags, among other specifics.
+        pktLine(sb, COMMAND_FETCH + '=' + OPTION_WAIT_FOR_DONE + ' ' + OPTION_SHALLOW);
+        // TODO(minwoox): Migrate hash function https://git-scm.com/docs/hash-function-transition
+        pktLine(sb, "object-format=sha1");
+        pktFlush(sb);
+        return sb.toString();
+    }
+
+    // https://git-scm.com/docs/protocol-common#_pkt_line_format
+    static void pktLine(StringBuilder sb, String line) {
+        lineLength(sb, line.length() + 5);
```

Comment:
Yeah, we can do that. 👍

Reply:
Let me revert this because we are using it for non-binary data. We need to use the length of the string rather than the length of the bytes. Although we are only using 1-byte characters, the outcomes may not differ; however, I want to avoid confusing others who read this code.

Reply:
Here are the references for non-binary data.
https://git-scm.com/docs/protocol-common#_pkt_line_format
https://lore.kernel.org/all/20210428222219.GA982@localhost/T/
```
As a general rule of thumb, if the pkt-line is a pack or object, it's
binary, and otherwise (a ref advertisement, a negotiation, etc.) it's
text.  Almost any case where GIT_TRACE_PACKET prints the output is going
to be text, and almost any case where it does not is going to be binary.
```

Reply:
I see, the reasoning I originally suggested this was:
1) I didn't think any protocols will use utf-16 code units for the length
2) The jGit library seems to use utf-8 as well
Having said this, I think it's fine if <=127 characters are used only.

Reply:
Oops, I was totally wrong about this. Let me revert the revert again. 😅